subworkflows/local/align_cellranger.nf

```diff
@@ -2,36 +2,46 @@
  * Alignment with Cellranger
  */
 
-params.cellranger_mkgtf_options    = [:]
-params.cellranger_mkref_options    = [:]
-params.cellranger_count_options    = [:]
-
-include {METADATA} from "../../modules/local/software/metadata/main.nf"
-include {CELLRANGER_MKGTF} from "../../modules/local/software/cellranger/mkgtf/main.nf"
-include {CELLRANGER_MKREF} from "../../modules/local/software/cellranger/mkref/main.nf"
-include {CELLRANGER_COUNT} from "../../modules/local/software/cellranger/count/main.nf"
+include {CELLRANGER_MKGTF} from "../../modules/nf-core/modules/cellranger/mkgtf/main.nf"
+include {CELLRANGER_MKREF} from "../../modules/nf-core/modules/cellranger/mkref/main.nf"
+include {CELLRANGER_COUNT} from "../../modules/nf-core/modules/cellranger/count/main.nf"
 
 // Define workflow to subset and index a genome region fasta file
 workflow CELLRANGER_ALIGN {
     take:
         fasta
         gtf
-        samplesheet
+        cellranger_index
+        ch_fastq
 
     main:
-        // Set sample channel from samplesheet input
-        METADATA            ( samplesheet )
+        ch_versions = Channel.empty()
+
+        assert cellranger_index || (fasta && gtf):
+            "Must provide either a cellranger index or both a fasta file ('--genome_fasta') and a gtf file ('--gtf')."
 
-        // Filter GTF based on gene biotypes passed in params.modules
-        CELLRANGER_MKGTF    ( gtf )
+        if (!cellranger_index) {
+            // Filter GTF based on gene biotypes passed in params.modules
+            CELLRANGER_MKGTF( gtf )
+            ch_versions = ch_versions.mix(CELLRANGER_MKGTF.out.versions)
 
-        // Make reference genome
-        CELLRANGER_MKREF    ( CELLRANGER_MKGTF.out.gtf, fasta )
+            CELLRANGER_MKGTF.out.gtf.view()
+            // Make reference genome
+            CELLRANGER_MKREF( fasta, CELLRANGER_MKGTF.out.gtf, "cellranger_reference" )
+            ch_versions = ch_versions.mix(CELLRANGER_MKREF.out.versions)
+            cellranger_index = CELLRANGER_MKREF.out.reference
+        }
 
         // Obtain read counts
-        CELLRANGER_COUNT    ( METADATA.out, CELLRANGER_MKREF.out.reference_genome )
+        CELLRANGER_COUNT (
+             // TODO what is `gem` and why is it needed?
+             // needed to add .flatten() on reads as cellranger count module expects a flattened array
+             ch_fastq.map{ meta, reads -> [meta + ["gem": meta.id, "samples": [meta.id]], reads.flatten()] },
```

Comment:
Isn't this `.flatten()` here redundant with the `.flatten()` in `input_check.nf`? 

If so, I would indeed prefer the `.flatten()` here (and in the kallisto call) over having it in `input_check`, to keep the pairing information intact as long as possible.

Reply:
Yes! Nicely spotted. I did this one first before changing on the subworkflow, and forgot about it!

😄

Reply:
You prefer having them on the modules? I mean, this flatten is just to revert what happenned when applying groupTuple so it returns again to the normal convetions from nf-core modules.

That's why I thought to put it in `input_check` to make sure what comes from there tries to follow adopted conventions. And also, having only on .flatten call, instead of multiple across modules :)

Reply:
Agree on removing it from there and leaving on `input_check`?

Reply:
Maybe add a comment in the module itself or `meta.yml` in what order multiple replicates are expected for the `reads` argument?

Reply:
In fact, this `.flatten()` would be required in all the modules where we have:

```groovy
input:
tuple val(meta), path(reads)
```

Because it only seems to work when this "reads" channel is a single array.

Reply:
> Maybe add a comment in the module itself or `meta.yml` in what order multiple replicates are expected for the `reads` argument?

so users always put their replicates properly ordered when creating the samplesheet, right?

Reply:
For now, I just removed the redundancy spotted. If we actually decide on going the other way, and having it on the modules, it does not seem like a complex change to make.

And about the comment, would we have to make these comments directly on our branch? Or have it on "nf-core/modules"? I ask because changing these files on our branch only may hamper linting stuff and future update of modules using nf-core tool, doesn't it?

Reply:
> so users always put their replicates properly ordered when creating the samplesheet, right?

you mean, always putting the corresponding pair of fastqs in one row? I would hope so! 
If the files follow some naming convention, we could technically try to validate it, but if now there's not a lot we can do about it anyway...

> For now, I just removed the redundancy spotted. If we actually decide on going the other way, and having it on the modules, it does not seem like a complex change to make.

:ok_hand: perfect

> And about the comment, would we have to make these comments directly on our branch?

I only meant the `local` modules for now. The linter shouldn't care about them.

Reply:
Was thinking in adding a comment like this in the modules?

```nextflow
//
// Input reads are expected to come as: [ meta, [ pair1_read1, pair1_read2, pair2_read1, pair2_read2 ] ]
// The input array for a sample is created in the same order reads appear in samplesheet as each pair of reads is appended to the array when its line is processed.
//
```

If so I can already add it to the relevant modules.
Then we can later discuss what's still missing for the PR 😄

Reply:
Added on our relevant local files:

* `star_align.nf`
* `kallistobustools_count.nf`
* `salmon_alevin.nf`